File: src/pt/up/fe/luisfonseca/cp/api/WarningsLoader.java
package pt.up.fe.luisfonseca.cp.api;

import pt.up.fe.luisfonseca.cp.api.json.Warnings;

import com.google.gson.Gson;

public class WarningsLoader extends RetrieveTask<Warnings> {
	
	private static class WarningsParser implements JsonParser<Warnings>
	{
		public Warnings parse(String json) {
	        Gson gson = new Gson();
	        
			return gson.fromJson(json, Warnings.class);
		}
	}

	public WarningsLoader(ResponseHandler<Warnings> handler) {
		super(handler, new WarningsParser());
		this.execute(CP.URLwarnings);
	}
}

File: src/pt/up/fe/luisfonseca/cp/api/WarningsLoader.scala
package pt.up.fe.luisfonseca.cp.api

import org.json.JSONObject
import scala.util.parsing.json.JSON
import pt.up.fe.luisfonseca.cp.api.json.{Warnings, WarningItem}
import org.json.JSONArray

class WarningsLoader(val handler: ResponseHandler[Warnings])
	extends RetrieveTask[Warnings](handler, new WarningsParser())
{
	def execute = super.execute(CP.URLwarnings);
}

class WarningsParser extends JsonParser[Warnings]
{
	def parse(json: String) : Warnings = {
	  
	  val jObject = new JSONObject(json);
	  if(jObject.has("today") && jObject.has("tomorrow") && jObject.has("after"))
	  {  
		  def extractNewsItem(ja: JSONArray): List[WarningItem] = {
		    for(i<- (0 to ja.length() - 1).toList)
			  yield new WarningItem(ja.getJSONObject(i).getString("title"),
		    						ja.getJSONObject(i).getString("desc"))
		  }
		  
		  val today = extractNewsItem(jObject.getJSONArray("today"))
		  val tomorrow = extractNewsItem(jObject.getJSONArray("tomorrow"))
		  val after = extractNewsItem(jObject.getJSONArray("after"))

		  new Warnings(today, tomorrow, after)
	  }
	  else
	    null
	}
}